### PR TITLE
Updating checksum middlewares and related API docs for S3/Glacier

### PR DESCRIPTION
**Pull request by jeremeamia**
Inspired by #655, this PR aims to do the following:

* Support `ContentSHA256` as a documented parameter for S3 and Glacier upload operations.
* Remove `ContentMD5` from API documentation for S3 upload operations (still works, just doesn't show up in docs).
* Support a non-capitalized `contentSHA256` in Glacier to match other params (capitalized still works).
* Fixes a bug in the Glacier middleware where the operation name for `UploadMultipartPart` was incorrect.
* Fixes the API documentation for Glacier upload operations for the `sourceFile` param. Was previously listed with a capital "S".
* Updates the API docs for all Glacier operations such that `accountId` is no longer marked as required.

@mtdowling and @jeskew, please take a look and let me know if you have any more ideas about ways to optimize or improve usability around these items.

### CODE REVIEW

**Comment by jeremeamia** [comment]
@mtdowling and @jeskew ping

**Comment by jeskew** [comment]
LGTM.

File: src/Glacier/GlacierClient.php [hunk 1]
```diff
@@ -141,17 +147,29 @@ public static function applyDocFilters(array $api, array $docs)
...
-        $api['shapes']['UploadArchiveInput']['members']['SourceFile'] = ['shape' => 'SourceFile'];
-        $api['shapes']['UploadMultipartPartInput']['members']['SourceFile'] = ['shape' => 'SourceFile'];
+        $api['shapes']['UploadArchiveInput']['members']['sourceFile'] = ['shape' => 'SourceFile'];
+        $api['shapes']['UploadMultipartPartInput']['members']['sourceFile'] = ['shape' => 'SourceFile'];
 
-        // Add information about "checksum" being optional.
-        $docs['shapes']['checksum']['append'] =
-            '<div class="alert alert-info">The SDK will compute this value '
+        // Add the ContentSHA256 parameter.
+        $docs['shapes']['ContentSHA256']['base'] = 'A SHA256 hash of the content of the request body';
```
Comment (mtdowling) [inline src/Glacier/GlacierClient.php]: Can you mention that this is calculated automatically? Most users will and should ignore this parameter so I want to make sure the docs are clear.

Comment (jeremeamia) [inline src/Glacier/GlacierClient.php]: Happens a few lines down.

Comment (mtdowling) [inline src/Glacier/GlacierClient.php]: Ah I see

File: tests/S3/ApplyChecksumMiddlewareTest.php [hunk 2]
```diff
@@ -61,4 +61,50 @@ public function getContentMd5UseCases()
...
+        );
+        $s3->execute($command);
+    }
+
+    public function getContentSha256UseCases()
+    {
+        $hash = 'SHA256HASH';
+
+        return [
+            // Do nothing if ContentSHA256 was not provided.
```
Comment (dstevenson) [inline tests/S3/ApplyChecksumMiddlewareTest.php]: @jeremeamia The `S3SignatureV4` will always try to set the `x-amz-content-sha256` header to the hash of the request body or POST fields on a `PutObject`.

Comment (jeremeamia) [inline tests/S3/ApplyChecksumMiddlewareTest.php]: Yep, the changes to this middleware make it so the user can provide the `x-amz-content-sha256` header. The Signature checks for the presence of the header before doing the work to calculate the signature. Setting the header manually is basically on optimization, or in the case of a non-seekable stream, a necessity.

Comment (dstevenson) [inline tests/S3/ApplyChecksumMiddlewareTest.php]: Ah, sorry, didn't see this test was appending the build step not the sign step

Comment (jeremeamia) [inline tests/S3/ApplyChecksumMiddlewareTest.php]: No worries! We appreciate you taking a look at the PR. :-)

**Pull request merged**